service/src/main/java/com/epam/esm/service/service/certificate/CertificateServiceImpl.java

```diff
@@ -0,0 +1,247 @@
+import com.epam.esm.persistence.util.finder.impl.CertificateFinder;
+import com.epam.esm.persistence.util.finder.impl.TagFinder;
+import com.epam.esm.service.constants.CertificateSearchParameters;
+import com.epam.esm.service.constants.CertificateSortingParameters;
+import com.epam.esm.service.constants.ErrorCodes;
+import com.epam.esm.service.constants.PaginationParameters;
+import com.epam.esm.service.exceptions.BadRequestException;
+import com.epam.esm.service.exceptions.NotFoundException;
+import com.epam.esm.service.exceptions.ValidationException;
+import com.epam.esm.service.service.tag.TagService;
+import com.epam.esm.service.validator.EntityValidator;
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Lookup;
+import org.springframework.dao.DataIntegrityViolationException;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+import org.springframework.util.MultiValueMap;
+
+import java.math.BigDecimal;
+import java.time.LocalDateTime;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Consumer;
+
+/**
+ * {@link Certificate} service class.
+ *
+ * @author Yury Zmushko
+ * @version 1.0
+ */
+@Service
+public class CertificateServiceImpl implements CertificateService {
+
+    private final CertificateDAO dao;
+    private final EntityValidator<Certificate> validator;
+    private final TagService tagService;
+
+    @Autowired
+    public CertificateServiceImpl(CertificateDAO dao,
+                                  EntityValidator<Certificate> validator,
+                                  TagService tagService) {
+        this.dao = dao;
+        this.validator = validator;
+        this.tagService = tagService;
+    }
+
+    @Transactional
+    @Override
+    public Certificate create(Certificate certificate)
+            throws ValidationException, BadRequestException {
+
+        certificate.setCreateDate(LocalDateTime.now());
+        certificate.setLastUpdateDate(LocalDateTime.now());
+        validator.validate(certificate);
+        for (Tag tag : certificate.getTags()) {
+            if (tagService.readOptional(tag.getId()).isEmpty()) {
+                tagService.create(tag);
+            }
+        }
+        try {
+        return dao.create(certificate);
+        } catch (DataIntegrityViolationException e) {
+            throw new BadRequestException(e, ErrorCodes.CERTIFICATE_BAD_REQUEST);
+        }
+    }
+
+    @Override
+    public Certificate read(int id) throws NotFoundException {
+        Optional<Certificate> certificateOptional = readOptional(id);
+        if (certificateOptional.isEmpty()) {
+            throw new NotFoundException("Requested resource not found(id = " + id + ")!",
+                    ErrorCodes.CERTIFICATE_NOT_FOUND);
+        } else {
```

Comment:
remove else, 
 just 'return' will be enought without else.

Reply:
Done